File: ports/go/sqlsec/sqlsec.go
// Package sqlsec is a Go port of the core sqlsec check: a passive,
// offline heuristic that flags unsafe SQL-query construction in source text.
//
// It mirrors the highest-signal rules from the Python reference
// implementation (SQL001 concatenation, SQL002 f/template interpolation,
// SQL003 printf-style formatting) so the same finding ids surface across
// language ecosystems. Defensive / educational scope only; it never executes
// anything and makes no network calls.
//
// Original work by Cognis Digital.
package sqlsec

import (
	"regexp"
	"strings"
)

// Severity levels, matching the reference tool's vocabulary.
type Severity string

const (
	Low      Severity = "low"
	Medium   Severity = "medium"
	High     Severity = "high"
	Critical Severity = "critical"
)

// Finding is one detected unsafe pattern at a 1-based line number.
type Finding struct {
	RuleID   string   `json:"rule_id"`
	Severity Severity `json:"severity"`
	Line     int      `json:"line"`
	Message  string   `json:"message"`
	Snippet  string   `json:"snippet"`
}

var sqlKeywords = regexp.MustCompile(
	`(?i)\b(SELECT|INSERT\s+INTO|UPDATE|DELETE\s+FROM|DROP\s+TABLE|` +
		`CREATE\s+TABLE|ALTER\s+TABLE|FROM|WHERE|VALUES|JOIN|UNION|` +
		`TRUNCATE|GRANT|REVOKE)\b`)

// looksLikeSQL reports whether a line plausibly contains a SQL fragment.
func looksLikeSQL(line string) bool {
	return sqlKeywords.MatchString(line)
}

type rule struct {
	id       string
	severity Severity
	pattern  *regexp.Regexp
	message  string
	needsSQL bool
}

var rules = []rule{
	{
		id:       "SQL001",
		severity: High,
		pattern:  regexp.MustCompile(`["'].*?["']\s*\+\s*\w|\w\s*\+\s*["'].*?["']`),
		message:  "SQL query built by string concatenation with a variable",
		needsSQL: true,
	},
	{
		// Go-style fmt.Sprintf / template interpolation of a value into SQL.
		id:       "SQL002",
		severity: High,
		pattern:  regexp.MustCompile(`(?i)(Sprintf|Sprint)\s*\(\s*["'].*?(%[sdvq]|\{).*?["']`),
		message:  "value interpolated into a SQL statement via Sprintf/format",
		needsSQL: true,
	},
	{
		id:       "SQL003",
		severity: High,
		pattern:  regexp.MustCompile(`["'].*?%[sd].*?["']\s*%\s*[\w(]`),
		message:  "SQL string assembled with printf-style formatting",
		needsSQL: true,
	},
}

// ScanText applies the rule set to source text line by line.
func ScanText(text string) []Finding {
	var findings []Finding
	for i, raw := range strings.Split(text, "\n") {
		line := strings.TrimRight(raw, "\r")
		if strings.TrimSpace(line) == "" {
			continue
		}
		for _, r := range rules {
			if r.needsSQL && !looksLikeSQL(line) {
				continue
			}
			if r.pattern.MatchString(line) {
				findings = append(findings, Finding{
					RuleID:   r.id,
					Severity: r.severity,
					Line:     i + 1,
					Message:  r.message,
					Snippet:  strings.TrimSpace(line),
				})
			}
		}
	}
	return findings
}

File: ports/ts/src/sqlsec.ts
/**
 * TypeScript port of the core sqlsec check: a passive, offline heuristic that
 * flags unsafe SQL-query construction in source text.
 *
 * Mirrors the high-signal rules from the Python reference (SQL001
 * concatenation, SQL002 template-literal interpolation, SQL003 printf-style
 * formatting) so finding ids stay consistent across language ecosystems.
 * Defensive / educational scope only; executes nothing, no network.
 *
 * Original work by Cognis Digital.
 */

export type Severity = "low" | "medium" | "high" | "critical";

export interface Finding {
  ruleId: string;
  severity: Severity;
  line: number; // 1-based
  message: string;
  snippet: string;
}

const SQL_KEYWORDS =
  /\b(SELECT|INSERT\s+INTO|UPDATE|DELETE\s+FROM|DROP\s+TABLE|CREATE\s+TABLE|ALTER\s+TABLE|FROM|WHERE|VALUES|JOIN|UNION|TRUNCATE|GRANT|REVOKE)\b/i;

function looksLikeSql(line: string): boolean {
  return SQL_KEYWORDS.test(line);
}

interface Rule {
  id: string;
  severity: Severity;
  pattern: RegExp;
  message: string;
  needsSql: boolean;
}

const RULES: Rule[] = [
  {
    id: "SQL001",
    severity: "high",
    pattern: /["'].*?["']\s*\+\s*\w|\w\s*\+\s*["'].*?["']/,
    message: "SQL query built by string concatenation with a variable",
    needsSql: true,
  },
  {
    // Template literal interpolation: `... ${value} ...` containing SQL.
    id: "SQL002",
    severity: "high",
    pattern: /`[^`]*\$\{[^}]+\}[^`]*`/,
    message: "value interpolated into a SQL statement via template literal",
    needsSql: true,
  },
  {
    id: "SQL003",
    severity: "high",
    pattern: /["'].*?%[sd].*?["']\s*%\s*[\w(]/,
    message: "SQL string assembled with printf-style formatting",
    needsSql: true,
  },
];

/** Apply the rule set to source text, returning findings line by line. */
export function scanText(text: string): Finding[] {
  const findings: Finding[] = [];
  const lines = text.split("\n");
  lines.forEach((raw, i) => {
    const line = raw.replace(/\r$/, "");
    if (line.trim() === "") return;
    for (const r of RULES) {
      if (r.needsSql && !looksLikeSql(line)) continue;
      if (r.pattern.test(line)) {
        findings.push({
          ruleId: r.id,
          severity: r.severity,
          line: i + 1,
          message: r.message,
          snippet: line.trim(),
        });
      }
    }
  });
  return findings;
}
